File: app/src/main/java/com/kickstarter/services/apiresponses/ShippingRulesEnvelope.java
package com.kickstarter.services.apiresponses;

import android.os.Parcelable;

import com.kickstarter.libs.qualifiers.AutoGson;
import com.kickstarter.models.ShippingRule;

import java.util.List;

import auto.parcel.AutoParcel;

@AutoGson
@AutoParcel
public abstract class ShippingRulesEnvelope implements Parcelable {
  public abstract List<ShippingRule> shippingRules();

  @AutoParcel.Builder
  public abstract static class Builder {
    public abstract Builder shippingRules(List<ShippingRule> ___);
    public abstract ShippingRulesEnvelope build();
  }

  public static Builder builder() {
    return new AutoParcel_ShippingRulesEnvelope.Builder();
  }

  public abstract Builder toBuilder();
}

File: app/src/main/java/com/kickstarter/services/apiresponses/ShippingRulesEnvelope.kt
package com.kickstarter.services.apiresponses

import android.os.Parcelable
import com.kickstarter.models.ShippingRule
import kotlinx.parcelize.Parcelize

@Parcelize
class ShippingRulesEnvelope private constructor(
    private val shippingRules: List<ShippingRule>
) : Parcelable {
    fun shippingRules() = this.shippingRules

    @Parcelize
    data class Builder(
        private var shippingRules: List<ShippingRule> = emptyList()
    ) : Parcelable {
        fun shippingRules(shippingRules: List<ShippingRule>) = apply { this.shippingRules = shippingRules }
        fun build() = ShippingRulesEnvelope(
            shippingRules = shippingRules
        )
    }

    fun toBuilder() = Builder(
        shippingRules = shippingRules
    )

    companion object {
        @JvmStatic
        fun builder() = Builder()
    }

    override fun equals(obj: Any?): Boolean {
        var equals = super.equals(obj)
        if (obj is ShippingRulesEnvelope) {
            equals = shippingRules() == obj.shippingRules()
        }
        return equals
    }
}
